src/Symfony/Component/Security/Http/MagicLink/MagicLoginLinker.php

```diff
@@ -0,0 +1,122 @@
+use Symfony\Component\Security\Core\User\UserInterface;
+use Symfony\Component\Security\Http\MagicLink\Exception\ExpiredMagicLinkException;
+use Symfony\Component\Security\Http\MagicLink\Exception\InvalidMagicLinkException;
+
+/**
+ * @author Ryan Weaver <ryan@symfonycasts.com>
+ */
+class MagicLoginLinker implements MagicLoginLinkerInterface
+{
+    private const SELECTOR_LENGTH = 20;
+    private const QUERY_PARAM = 'token';
+
+    private $storage;
+    private $secret;
+    private $urlGenerator;
+    private $routeName;
+    private $lifetime;
+
+    public function __construct(MagicLinkTokenStorageInterface $storage, UrlGeneratorInterface $urlGenerator, string $secret, int $lifetime, string $routeName)
+    {
+        $this->storage = $storage;
+        $this->secret = $secret;
+        $this->urlGenerator = $urlGenerator;
+        $this->lifetime = $lifetime;
+        $this->routeName = $routeName;
+    }
+
+    public function createMagicLink(UserInterface $user): MagicLoginLinkDetails
+    {
+        $this->removeExpiredLinks();
+
+        // length will be 20
+        $selector = $this->generateRandomString(15);
+        $verifier = $this->generateRandomString(18);
+        $expiresAt = new \DateTimeImmutable(sprintf('+%d seconds', $this->lifetime));
+
+        $this->storage->storeToken(
+            $selector,
+            $this->hashVerifier($verifier),
+            $user,
+            $expiresAt
+        );
+
+        $publicToken = $selector.$verifier;
+        $url = $this->urlGenerator->generate(
+            $this->routeName,
+            [self::QUERY_PARAM => $publicToken],
+            UrlGeneratorInterface::ABSOLUTE_URL
+        );
+
+        return new MagicLoginLinkDetails($url, $expiresAt);
+    }
+
+    public function consumeMagicLink(string $uri): UserInterface
+    {
+        $this->removeExpiredLinks();
+
+        $requestToken = $this->getTokenFromUri($uri);
+
+        $selector = substr($requestToken, 0, self::SELECTOR_LENGTH);
+        $verifier = substr($requestToken, self::SELECTOR_LENGTH);
+        $hashedVerifier = $this->hashVerifier($verifier);
+
+        $token = $this->storage->findToken($selector);
+        // immediately invalidate, even if the verifier is wrong
+        $this->storage->invalidateToken($selector);
+
+        // always check a hash to not expose if a token was found by selector
+        $storedVerifier = $token ? $token->getHashedVerifier() : 'fake_verifier';
+
+        if (false === hash_equals($hashedVerifier, $storedVerifier)) {
+            throw new InvalidMagicLinkException('Invalid verifier.');
+        }
+
+        if ($token->getExpiresAt()->getTimestamp() <= time()) {
+            throw new ExpiredMagicLinkException('Magic link has expire.');
+        }
+
+        return $token->getUser();
+    }
+
+    private function generateRandomString(int $bytes): string
+    {
+        return strtr(base64_encode(random_bytes($bytes)), '+/', '-_');
```

Comment:
We could use `ByteString::fromRandom()` here, if depending on the String component is not a problem.

Reply:
Hmm, that's a nice function - I would love to use that.

What do others think? It would mean adding a small dependency to all of `symfony/security-http`.

Reply:
I'm not sure it really worth it 
But I'm not against it